File: 1249-minimum-remove-to-make-valid-parentheses/1249-minimum-remove-to-make-valid-parentheses.py
class Solution(object):
    def minRemoveToMakeValid(self, s):
        """
        :type s: str
        :rtype: str
        """
        stack = []
        idxTrash = set()
        ans = []
        
        for idx, char in enumerate(s):
            if char == '(':
                stack.append(idx)
            if char == ')':
                if stack:
                    stack.pop()
                else:
                    idxTrash.add(idx)
            
        while stack:
            idxTrash.add(stack.pop())
            
        for idx, char in enumerate(s):
            if idx not in idxTrash:
                ans.append(char)
        
        return ans

File: 1249-minimum-remove-to-make-valid-parentheses/1249-minimum-remove-to-make-valid-parentheses.java
class Solution {
    public String minRemoveToMakeValid(String s) {
        Set<Integer> idxTrash = new HashSet<>();
        Stack<Integer> stack = new Stack<>();
        StringBuilder sb = new StringBuilder();
        
        for (int i = 0; i < s.length(); i++) {
            if (s.charAt(i) == '(') {
                stack.push(i);
            }
            if (s.charAt(i) == ')') {
                if (!stack.isEmpty()) {
                    stack.pop();
                }
                else {idxTrash.add(i);}
            }
        }
        
        while (!stack.isEmpty()) {
            idxTrash.add(stack.pop());
        }
        
        for (int i = 0; i < s.length(); i++) {
            if (!idxTrash.contains(i)) {
                sb.append(s.charAt(i));
            }
        }
        
        return sb.toString();
    }
}